tests/posix/pthread_pressure/src/main.c

```diff
@@ -0,0 +1,223 @@
+/*
+ * Copyright (c) 2023, Meta
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <pthread.h>
+
+#include <zephyr/sys/util.h>
+#include <zephyr/ztest.h>
+
+#define STACK_SIZE K_THREAD_STACK_LEN(CONFIG_TEST_STACK_SIZE)
+
+/* update interval for printing stats */
+#if CONFIG_TEST_DURATION_S >= 60
+#define UPDATE_INTERVAL_S 10
+#elif CONFIG_TEST_DURATION_S >= 30
+#define UPDATE_INTERVAL_S 5
+#else
+#define UPDATE_INTERVAL_S 1
+#endif
+
+/* 32 threads is mainly a limitation of find_lsb_set() */
+#define NUM_THREADS MIN(32, MIN(CONFIG_TEST_NUM_CPUS, CONFIG_MAX_PTHREAD_COUNT))
+
+typedef int (*create_fn)(int i);
+typedef int (*join_fn)(int i);
+
+static void *setup(void);
+static void before(void *fixture);
+
+/* bitmask of available threads */
+static bool alive[NUM_THREADS];
+
+/* array of thread stacks */
+static K_THREAD_STACK_ARRAY_DEFINE(thread_stacks, NUM_THREADS, STACK_SIZE);
+
+static struct k_thread k_threads[NUM_THREADS];
+static size_t counters[NUM_THREADS];
+
+static void print_stats(uint64_t now, uint64_t end)
+{
+	printk("now (ms): %llu end (ms): %llu\n", now, end);
+	for (int i = 0; i < NUM_THREADS; ++i) {
+		printk("Thread %d created and joined %zu times\n", i, counters[i]);
+	}
+}
+
+static void test_create_join_common(const char *tag, create_fn create, join_fn join)
+{
+	int i;
+	__unused int ret;
+	uint64_t now_ms = k_uptime_get();
+	const uint64_t end_ms = now_ms + MSEC_PER_SEC * CONFIG_TEST_DURATION_S;
+	uint64_t update_ms = now_ms + MSEC_PER_SEC * UPDATE_INTERVAL_S;
+
+	printk("BOARD: %s\n", CONFIG_BOARD);
+	printk("NUM_THREADS: %u\n", NUM_THREADS);
+	printk("TEST_NUM_CPUS: %u\n", CONFIG_TEST_NUM_CPUS);
+	printk("TEST_DURATION_S: %u\n", CONFIG_TEST_DURATION_S);
+	printk("TEST_DELAY_US: %u\n", CONFIG_TEST_DELAY_US);
+
+	for (i = 0; i < NUM_THREADS; ++i) {
+		/* spawn thread i */
+		ret = create(i);
+#ifdef CONFIG_TEST_EXTRA_ASSERTIONS
+		zassert_ok(ret, "%s_create(%d)[%zu] failed: %d", tag, i, counters[i], ret);
+#endif
+	}
+
+	do {
+#ifndef CONFIG_SMP
+		/* allow the test thread to be swapped-out */
+		k_yield();
+#endif
+
+		for (i = 0; i < NUM_THREADS; ++i) {
+			if (alive[i]) {
+				ret = join(i);
+#ifdef CONFIG_TEST_EXTRA_ASSERTIONS
+				zassert_ok(ret, "%s_join(%d)[%zu] failed: %d", tag, i, counters[i],
+					   ret);
+#endif
+				alive[i] = false;
+
+				/* update counter i after each (create,join) pair */
+				++counters[i];
+
+#if CONFIG_TEST_DELAY_US > 0
+				/* success with 0 delay means we are ~raceless */
+				k_busy_wait(CONFIG_TEST_DELAY_US);
```

Comment:
Will a busy wait ever yield the CPU in a non-SMP system?  Would that be necessary to work around the k_thread race condition bug?

Reply:
It shouldn't yield the CPU even in SMP systems. I was just hoping to avoid conditionals on the hot path, since that seems to be a contributor to the sched assertions I've seen.

Reply:
yeah, as long as these tests pass and we have a reasonably reliable pattern for developers to use with the API things are good. Getting k_threads fixed can be a post 3.4 activity.